src/io/aleph/dirigiste/Pool.java

```diff
@@ -516,7 +518,21 @@ public void dispose(K key, V obj) {
         }
 
         if (pendingTakes > 0) {
-            addObject(key);
+            try {
+                // To prevent the queue from being deleted by the startControlLoop method (which runs
+                // on another thread) as soon as it has been created, we need to mark the Queue
+                // as in use after acquired the exclusive lock.
+                _lock.lock();
+                _lockedQueues.compute(key, (__, v) -> v == null ? 1 : v + 1);
+                _lock.unlock();
+                // Objects can be created when there are pending takes. Under this circumstance, a
+                // new object has to be created to replace the one that just got disposed.
+                addObject(key);
+            } finally {
+                // This call is safe, there is no reason the key won't be present on the locked
+                // queues at this point.
+                _lockedQueues.compute(key, (__, v) -> v - 1);
+            }
```

Comment:
I dislike this pattern. What about adding an AtomicBoolean or a volatile boolean flag to each Queue, called something like `canBeCleanedUp`. It starts off false, and then is set to true as soon as something is added. Might not even need to be volatile, honestly, since if it's out-of-date from the control loop thread's perspective, it just means it'll delay clean up a bit longer.

Or, we could change how Queue creation works, place newly-made Queues in a "nursery" that the control loop can't see, and only move them to the main `_queues` map once something's been added to them once. Which would probably be even more work, now that I think about it.

Part of how I think about this is, if using the global pool lock more often reduces performance, the value of Dirigiste is lowered, and it's already pretty low to me as is 😄 .

Reply:
> Might not even need to be volatile, honestly, since if it's out-of-date from the control loop thread's perspective, it just means it'll delay clean up a bit longer.

Apparently, this is not actually guaranteed, even if it's common. The JVM might never update it, or the behavior might be architecture-dependent.

Reply:
Well, my initial smoke test of this didn't work so hot, so maybe not. I think it would still work, but I don't want to spend much time hammering this into shape.

Reply:
> I dislike this pattern. What about adding an AtomicBoolean or a volatile boolean flag to each Queue, called something like canBeCleanedUp. It starts off false, and then is set to true as soon as something is added. Might not even need to be volatile, honestly, since if it's out-of-date from the control loop thread's perspective, it just means it'll delay clean up a bit longer.

This idea is not bad, but as soon as `canBeCleanedUp` changes from `false` to `true`, we need to be sure it can be collected once empty.

> Or, we could change how Queue creation works, place newly-made Queues in a "nursery" that the control loop can't see, and only move them to the main _queues map once something's been added to them once. Which would probably be even more work, now that I think about it.

I think you are missing one part of the problem.
Yes, at creation, it's a problem that could be solved by what you propose.
But... it's also a problem when it becomes empty and then got refilled in between when it could have been garbage collected.
That's why the `nursery` won't solve the issue in this case.

You make a good point though. I would love finding an approach that works that way. But I don't have it... yet.